clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java

```diff
@@ -117,29 +128,94 @@ public MockProducer() {
         this(Cluster.empty(), false, null, null, null);
     }
 
-    public void initTransactions() {
+    private <T> ExtendedSerializer<T> ensureExtended(Serializer<T> serializer) {
+        return serializer instanceof ExtendedSerializer ? (ExtendedSerializer<T>) serializer : new ExtendedSerializer.Wrapper<>(serializer);
+    }
 
+    @Override
+    public void initTransactions() {
+        verifyProducerState();
+        if (this.transactionInitialized) {
+            throw new IllegalStateException("MockedProducer got already initialized for transactions.");
+        }
+        this.transactionInitialized = true;
     }
 
+    @Override
     public void beginTransaction() throws ProducerFencedException {
-
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        this.transactionInFlight = true;
+        this.transactionCommitted = false;
+        this.transactionAborted = false;
     }
 
+    @Override
     public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
-                                  String consumerGroupId) throws ProducerFencedException {
-
+                                         String consumerGroupId) throws ProducerFencedException {
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        if (!transactionInFlight) {
+            throw new IllegalStateException("There is not open transaction to be committed.");
+        }
+        Map<TopicPartition, OffsetAndMetadata> uncommittedOffsets = this.uncommittedConsumerGroupOffsets.get(consumerGroupId);
+        if (uncommittedOffsets == null) {
+            uncommittedOffsets = new HashMap<>();
+            this.uncommittedConsumerGroupOffsets.put(consumerGroupId, uncommittedOffsets);
+        }
+        uncommittedOffsets.putAll(offsets);
     }
 
+    @Override
     public void commitTransaction() throws ProducerFencedException {
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        if (!transactionInFlight) {
+            throw new IllegalStateException("There is not open transaction to be committed.");
+        }
+
+        flush();
+
+        this.sent.addAll(this.uncommittedSends);
+        if (!this.uncommittedConsumerGroupOffsets.isEmpty())
+            this.consumerGroupOffsets.add(this.uncommittedConsumerGroupOffsets);
+
+        this.uncommittedSends.clear();
+        this.uncommittedConsumerGroupOffsets = new HashMap<>();
+        this.transactionCommitted = true;
+        this.transactionAborted = false;
+        this.transactionInFlight = false;
 
     }
 
+    @Override
     public void abortTransaction() throws ProducerFencedException {
-        
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        if (!transactionInFlight) {
```

Comment:
ok - same thing three times. Maybe extract it to a method `verifyTransactionInflight`